File: app/src/main/java/com/apps/adrcotfas/goodtime/Statistics/Main/SelectLabelDialog.java
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.apps.adrcotfas.goodtime.Statistics.Main;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;

import com.apps.adrcotfas.goodtime.Settings.PreferenceHelper;
import com.apps.adrcotfas.goodtime.Label;
import com.apps.adrcotfas.goodtime.AddEditLabels.AddEditLabelActivity;
import com.apps.adrcotfas.goodtime.Main.LabelsViewModel;
import com.apps.adrcotfas.goodtime.Profile;
import com.apps.adrcotfas.goodtime.R;
import com.apps.adrcotfas.goodtime.Settings.ProfilesViewModel;
import com.apps.adrcotfas.goodtime.Util.ThemeHelper;
import com.apps.adrcotfas.goodtime.databinding.DialogSelectLabelBinding;
import com.google.android.material.chip.Chip;


import java.lang.ref.WeakReference;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProvider;

import static com.apps.adrcotfas.goodtime.Statistics.Utils.getInstanceTotalLabel;
import static com.apps.adrcotfas.goodtime.Statistics.Utils.getInstanceUnlabeledLabel;
import static com.apps.adrcotfas.goodtime.Util.ThemeHelper.COLOR_INDEX_ALL_LABELS;
import static com.apps.adrcotfas.goodtime.Util.UpgradeActivityHelper.launchUpgradeActivity;

public class SelectLabelDialog extends DialogFragment {

    /**
     * The callback used to indicate the user is done selecting the title
     */
    public interface OnLabelSelectedListener {

        /**
         * @param label     the label that was set
         */
        void onLabelSelected(Label label);
    }

    private List<Profile> mProfiles;
    private String mLabel;
    private WeakReference<OnLabelSelectedListener> mCallback;
    /**
     * The extended version of this dialog is used in the Statistics
     * where it also contains "all" as a label.
     * The regular version contains an extra neutral button for selecting the current profile.
     */
    private boolean mIsExtendedVersion;
    private AlertDialog mAlertDialog;

    public SelectLabelDialog() {
        // Empty constructor required for DialogFragment
    }

    public static SelectLabelDialog newInstance(OnLabelSelectedListener listener, String label, boolean isExtendedVersion) {
        SelectLabelDialog dialog = new SelectLabelDialog();
        dialog.mCallback = new WeakReference<>(listener);
        dialog.mLabel = label;
        dialog.mIsExtendedVersion = isExtendedVersion;
        return dialog;
    }

    @SuppressLint("ResourceType")
    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        DialogSelectLabelBinding binding = DataBindingUtil.inflate(
                LayoutInflater.from(getContext()),
                R.layout.dialog_select_label,
                null,
                false);

        binding.editLabels.setOnClickListener(v -> {
            if (PreferenceHelper.isPro()) {
                Intent intent = new Intent(getActivity(), AddEditLabelActivity.class);
                startActivity(intent);
            } else {
                launchUpgradeActivity(getActivity());
            }

            if (mAlertDialog != null) {
                mAlertDialog.dismiss();
            }
        });

        LabelsViewModel labelsVm = new ViewModelProvider(this).get(LabelsViewModel.class);
        labelsVm.getLabels().observe(this, labels -> {

            int i = 0;
            if (mIsExtendedVersion) {
                Chip chip = new Chip(requireContext());
                Label total = getInstanceTotalLabel(requireContext());
                chip.setText(total.title);
                chip.setChipBackgroundColor(ColorStateList.valueOf(ThemeHelper.getColor(getActivity(), COLOR_INDEX_ALL_LABELS)));
                chip.setCheckable(true);
                chip.setChipIcon(getResources().getDrawable(R.drawable.ic_check_off));
                chip.setCheckedIcon(getResources().getDrawable(R.drawable.ic_check));

                chip.setId(i++);
                if (chip.getText().toString().equals(mLabel)) {
                    chip.setChecked(true);
                }
                binding.labels.addView(chip);
            }

            for (int j = labels.size() - 1; j >= 0; --j) {
                Label crt = labels.get(j);
                Chip chip = new Chip(requireContext());
                chip.setText(crt.title);
                chip.setChipBackgroundColor(ColorStateList.valueOf(ThemeHelper.getColor(getActivity(), crt.colorId)));
                chip.setCheckable(true);
                chip.setChipIcon(getResources().getDrawable(R.drawable.ic_check_off));
                chip.setCheckedIcon(getResources().getDrawable(R.drawable.ic_check));

                chip.setId(i++);
                if (crt.title.equals(mLabel)) {
                    chip.setChecked(true);
                }
                binding.labels.addView(chip);
            }

            final Handler handler = new Handler();
            handler.postDelayed(() -> {
                binding.progressBar.setVisibility(View.GONE);
                if (binding.labels.getChildCount() == 0) {
                    binding.emptyState.setVisibility(View.VISIBLE);
                } else {
                    binding.emptyState.setVisibility(View.GONE);
                    binding.labelsView.setVisibility(View.VISIBLE);
                }
            }, 100);
        });

        AlertDialog.Builder builder = new AlertDialog.Builder(requireContext())
                .setView(binding.getRoot())
                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                    if (binding.labels.getCheckedChipId() != -1) {
                        Chip chip = (Chip) (binding.labels.getChildAt(binding.labels.getCheckedChipId()));
                        mLabel = chip.getText().toString();
                        int color = chip.getChipBackgroundColor().getDefaultColor();
                        notifyLabelSelected(new Label(mLabel, ThemeHelper.getIndexOfColor(requireContext(), color)));
                    } else {
                        notifyLabelSelected(getInstanceUnlabeledLabel());
                    }
                    dismiss();
                })
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> {
                    if (dialog != null) {
                        dialog.dismiss();
                    }
                });
        if (!mIsExtendedVersion) {
            builder.setNeutralButton(
                    PreferenceHelper.isUnsavedProfileActive()
                            ? getResources().getString(R.string.Profile)
                            : PreferenceHelper.getProfile(), null);
            mAlertDialog = builder.create();
            mAlertDialog.setOnShowListener(dialog -> {

                //TODO: Clean-up this mess
                Button neutral = mAlertDialog.getButton(DialogInterface.BUTTON_NEUTRAL);
                neutral.setOnClickListener(v -> {
                    ProfilesViewModel profilesVm = new ViewModelProvider(SelectLabelDialog.this).get(ProfilesViewModel.class);
                    LiveData<List<Profile>> profilesLiveData = profilesVm.getProfiles();
                    profilesLiveData.observe(SelectLabelDialog.this, profiles -> {

                        mProfiles = profiles;
                        int profileIdx = 0;
                        final ArrayAdapter<String> arrayAdapter = new ArrayAdapter<>(requireContext(),
                                R.layout.checked_text_view);

                        String pref25 = SelectLabelDialog.this.getResources().getText(R.string.pref_profile_default).toString();
                        String pref52 = SelectLabelDialog.this.getResources().getText(R.string.pref_profile_5217).toString();
                        String crtProfileName = PreferenceHelper.getProfile();

                        if (crtProfileName.equals(pref25)) {
                            profileIdx = 0;
                        } else if (crtProfileName.equals(pref52)) {
                            profileIdx = 1;
                        }

                        arrayAdapter.add(pref25);
                        arrayAdapter.add(pref52);

                        final int PREDEFINED_PROFILES_NR = arrayAdapter.getCount();

                        for (int i = 0; i < profiles.size(); ++i) {
                            Profile p = profiles.get(i);
                            arrayAdapter.add(p.name);
                            if (crtProfileName.equals(p.name)) {
                                profileIdx = i + PREDEFINED_PROFILES_NR;
                            }
                        }

                        AlertDialog.Builder profileDialogBuilder =
                                new AlertDialog.Builder(requireContext())
                                        .setTitle(SelectLabelDialog.this.getResources().getString(R.string.Profile))
                                        .setSingleChoiceItems(
                                                arrayAdapter,
                                                PreferenceHelper.isUnsavedProfileActive() ? -1 : profileIdx,
                                                (dialogInterface, which) -> {
                                            String selected = arrayAdapter.getItem(which);
                                            updateProfile(which);

                                            dialogInterface.dismiss();
                                            if (mAlertDialog != null) {
                                                mAlertDialog.getButton(AlertDialog.BUTTON_NEUTRAL).setText(selected);
                                            }
                                        })
                                        .setNegativeButton(android.R.string.cancel, (dialog1, which) -> dialog1.dismiss());
                        profileDialogBuilder.show();
                    });
                });
            });
        } else {
            mAlertDialog = builder.create();
        }
        return mAlertDialog;
    }

    private void notifyLabelSelected(Label label) {
        if (mCallback != null) {
            mCallback.get().onLabelSelected(label);
        }
    }

    private void updateProfile(int index) {
        if (index == 0) {
            PreferenceHelper.setProfile25_5();
        } else if (index == 1) {
            PreferenceHelper.setProfile52_17();
        } else {
            int PREDEFINED_PROFILES_NR = 2;
            PreferenceHelper.setProfile(mProfiles.get(index - PREDEFINED_PROFILES_NR));
        }
    }
}

File: app/src/main/java/com/apps/adrcotfas/goodtime/statistics/main/SelectLabelDialog.kt
/*
 * Copyright 2016-2019 Adrian Cotfas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
package com.apps.adrcotfas.goodtime.statistics.main

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.DialogInterface
import android.content.Intent
import android.content.res.ColorStateList
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import androidx.appcompat.app.AlertDialog
import androidx.core.content.res.ResourcesCompat
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.viewModels
import com.apps.adrcotfas.goodtime.R
import com.apps.adrcotfas.goodtime.database.Label
import com.apps.adrcotfas.goodtime.database.Profile
import com.apps.adrcotfas.goodtime.databinding.DialogSelectLabelBinding
import com.apps.adrcotfas.goodtime.labels.AddEditLabelActivity
import com.apps.adrcotfas.goodtime.main.LabelsViewModel
import com.apps.adrcotfas.goodtime.settings.PreferenceHelper
import com.apps.adrcotfas.goodtime.settings.ProfilesViewModel
import com.apps.adrcotfas.goodtime.statistics.Utils
import com.apps.adrcotfas.goodtime.util.ThemeHelper
import com.apps.adrcotfas.goodtime.util.UpgradeDialogHelper.Companion.launchUpgradeDialog
import com.google.android.material.chip.Chip
import dagger.hilt.android.AndroidEntryPoint
import java.lang.ref.WeakReference
import javax.inject.Inject

@AndroidEntryPoint
class SelectLabelDialog : DialogFragment() {

    /**
     * The callback used to indicate the user is done selecting the title
     */
    interface OnLabelSelectedListener {
        /**
         * @param label     the label that was set
         */
        fun onLabelSelected(label: Label)
    }

    @Inject
    lateinit var preferenceHelper: PreferenceHelper

    private lateinit var mProfiles: List<Profile>
    private lateinit var mLabel: String
    private lateinit var mCallback: WeakReference<OnLabelSelectedListener>

    /**
     * The extended version of this dialog is used in the Statistics
     * where it also contains "all" as a label.
     */
    private var mIsExtendedVersion = false

    /**
     * The neutral button used to change the profile
     */
    private var showProfileSelection = false

    private var mAlertDialog: AlertDialog? = null

    @SuppressLint("ResourceType")
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val binding: DialogSelectLabelBinding = DataBindingUtil.inflate(
            layoutInflater,
            R.layout.dialog_select_label,
            null,
            false
        )
        binding.editLabels.setOnClickListener {
            if (preferenceHelper.isPro()) {
                val intent = Intent(activity, AddEditLabelActivity::class.java)
                startActivity(intent)
            } else {
                launchUpgradeDialog(requireActivity().supportFragmentManager)
            }
            if (mAlertDialog != null) {
                mAlertDialog!!.dismiss()
            }
        }
        val viewModel: LabelsViewModel by viewModels()
        viewModel.labels.observe(this) { labels: List<Label> ->
            var i = 0
            if (mIsExtendedVersion) {
                val chip = Chip(requireContext())
                val total = Utils.getInstanceTotalLabel(requireContext())
                chip.text = total.title
                chip.chipBackgroundColor = ColorStateList.valueOf(
                    ThemeHelper.getColor(
                        requireContext(),
                        ThemeHelper.COLOR_INDEX_ALL_LABELS
                    )
                )
                chip.isCheckable = true
                chip.chipIcon =
                    ResourcesCompat.getDrawable(resources, R.drawable.ic_check_off, null)
                chip.checkedIcon = ResourcesCompat.getDrawable(resources, R.drawable.ic_check, null)
                chip.id = i++
                if (chip.text.toString() == mLabel) {
                    chip.isChecked = true
                }
                binding.labels.addView(chip)
            }
            for (j in labels.indices.reversed()) {
                val crt = labels[j]
                val chip = Chip(requireContext())
                chip.text = crt.title
                chip.chipBackgroundColor =
                    ColorStateList.valueOf(ThemeHelper.getColor(requireContext(), crt.colorId))
                chip.isCheckable = true
                chip.chipIcon = ResourcesCompat.getDrawable(resources,R.drawable.ic_check_off, null)
                chip.checkedIcon = ResourcesCompat.getDrawable(resources,R.drawable.ic_check, null)
                chip.id = i++
                if (crt.title == mLabel) {
                    chip.isChecked = true
                }
                binding.labels.addView(chip)
            }
            if (binding.labels.childCount == 0) {
                binding.emptyState.visibility = View.VISIBLE
            } else {
                binding.emptyState.visibility = View.GONE
                binding.labelsView.visibility = View.VISIBLE
            }
        }
        val builder = AlertDialog.Builder(requireContext())
            .setView(binding.root)
            .setPositiveButton(android.R.string.ok) { _: DialogInterface?, _: Int ->
                if (binding.labels.checkedChipId != -1) {
                    val chip = binding.labels.getChildAt(binding.labels.checkedChipId) as Chip
                    mLabel = chip.text.toString()
                    val color = chip.chipBackgroundColor!!.defaultColor
                    notifyLabelSelected(
                        Label(
                            mLabel,
                            ThemeHelper.getIndexOfColor(requireContext(), color)
                        )
                    )
                } else {
                    notifyLabelSelected(Utils.getInstanceUnlabeledLabel(requireContext()))
                }
                dismiss()
            }
            .setNegativeButton(android.R.string.cancel) { dialog: DialogInterface?, _: Int -> dialog?.dismiss() }
        if (showProfileSelection) {
            builder.setNeutralButton(
                if (preferenceHelper.isUnsavedProfileActive()) resources.getString(R.string.Profile) else preferenceHelper.profile,
                null
            )
            mAlertDialog = builder.create()
            mAlertDialog!!.setOnShowListener {

                //TODO: Clean-up this mess
                val neutral = mAlertDialog!!.getButton(DialogInterface.BUTTON_NEUTRAL)
                neutral.setOnClickListener {
                    val profilesViewModel: ProfilesViewModel by viewModels()
                    val profilesLiveData = profilesViewModel.profiles
                    profilesLiveData.observe(this@SelectLabelDialog) { profiles: List<Profile> ->
                        mProfiles = profiles
                        var profileIdx = 0
                        val arrayAdapter = ArrayAdapter<String>(
                            requireContext(),
                            R.layout.checked_text_view
                        )
                        val pref25 =
                            this@SelectLabelDialog.resources.getText(R.string.pref_profile_default)
                                .toString()
                        val pref52 =
                            this@SelectLabelDialog.resources.getText(R.string.pref_profile_5217)
                                .toString()
                        val crtProfileName = preferenceHelper.profile
                        if (crtProfileName == pref25) {
                            profileIdx = 0
                        } else if (crtProfileName == pref52) {
                            profileIdx = 1
                        }
                        arrayAdapter.add(pref25)
                        arrayAdapter.add(pref52)
                        val predefinedProfileNum = arrayAdapter.count
                        for (i in profiles.indices) {
                            val p = profiles[i]
                            arrayAdapter.add(p.name)
                            if (crtProfileName == p.name) {
                                profileIdx = i + predefinedProfileNum
                            }
                        }
                        val profileDialogBuilder = AlertDialog.Builder(requireContext())
                            .setTitle(this@SelectLabelDialog.resources.getString(R.string.Profile))
                            .setSingleChoiceItems(
                                arrayAdapter,
                                if (preferenceHelper.isUnsavedProfileActive()) -1 else profileIdx
                            ) { dialogInterface: DialogInterface, which: Int ->
                                val selected = arrayAdapter.getItem(which)
                                updateProfile(which)
                                dialogInterface.dismiss()
                                if (mAlertDialog != null) {
                                    mAlertDialog!!.getButton(AlertDialog.BUTTON_NEUTRAL).text =
                                        selected
                                }
                            }
                            .setNegativeButton(android.R.string.cancel) { dialog1: DialogInterface, _: Int -> dialog1.dismiss() }
                        profileDialogBuilder.show()
                    }
                }
            }
        } else {
            mAlertDialog = builder.create()
        }
        return mAlertDialog!!
    }

    private fun notifyLabelSelected(label: Label) {
        mCallback.get()!!.onLabelSelected(label)
    }

    private fun updateProfile(index: Int) {
        when (index) {
            0 -> {
                preferenceHelper.setProfile25to5()
            }
            1 -> {
                preferenceHelper.setProfile52to17()
            }
            else -> {

                preferenceHelper.setProfile(mProfiles[index - PREDEFINED_PROFILES_NR])
            }
        }
    }

    companion object {
        const val PREDEFINED_PROFILES_NR = 2

        @JvmStatic
        fun newInstance(
            listener: OnLabelSelectedListener,
            label: String,
            isExtendedVersion: Boolean,
            showProfileSelection: Boolean = false
        ): SelectLabelDialog {
            val dialog = SelectLabelDialog()
            dialog.mCallback = WeakReference(listener)
            dialog.mLabel = label
            dialog.mIsExtendedVersion = isExtendedVersion
            dialog.showProfileSelection = showProfileSelection
            return dialog
        }
    }
}